app/src/org/commcare/android/database/app/models/InstanceRecord.java

```diff
@@ -0,0 +1,448 @@
+     */
+    private static String getDisplaySubtext(@InstanceStatus String status) {
+        String ts = new SimpleDateFormat("EEE, MMM dd, yyyy 'at' HH:mm").format(new Date());
+
+        if (status == null) {
+            return "Added on " + ts;
+        } else if (STATUS_INCOMPLETE.equalsIgnoreCase(status)) {
+            return "Saved on " + ts;
+        } else if (STATUS_COMPLETE.equalsIgnoreCase(status)) {
+            return "Finalized on " + ts;
+        } else if (STATUS_SUBMITTED.equalsIgnoreCase(status)) {
+            return "Sent on " + ts;
+        } else if (STATUS_SUBMISSION_FAILED.equalsIgnoreCase(status)) {
+            return "Sending failed on " + ts;
+        } else {
+            return "Added on " + ts;
+        }
+    }
+
+    public void delete() {
+        String instanceDir = (new File(mInstanceFilePath)).getParent();
+        deleteFileOrDir(instanceDir);
+        getInstanceRecordStorage().remove(getID());
+    }
+
+    public void save(@InstanceInsertType int insertType) {
+        getInstanceRecordStorage().write(this);
+
+        switch (insertType) {
+            case INSERTION_TYPE_SESSION_LINKED:
+                finalizeSessionLinkedInsertion();
+                break;
+            case INSERTION_TYPE_UNINDEXED_IMPORT:
+                finalizeUnindexedInsertion();
+                break;
+            case INSERTION_TYPE_SANDBOX_MIGRATED:
+                break;
+        }
+    }
+
+    private void finalizeSessionLinkedInsertion() {
+        // If we've changed a particular form instance's status, we need to mirror the
+        // change in that form's record.
+        try {
+            linkToSessionFormRecord();
+        } catch (IllegalStateException e) {
+            throw e;
+        } catch (Exception e) {
+            Logger.exception(e);
+            throw new SQLException("Failed to update Instance row " + getID());
+        }
+    }
+
+    private void finalizeUnindexedInsertion() {
+        SecretKey key = CommCareApplication.instance().createNewSymmetricKey();
+        FormRecord r = new FormRecord(getID(), FormRecord.STATUS_UNINDEXED,
+                mJrFormId, key.getEncoded(), null, new Date(0), ProviderUtils.getSandboxedAppId());
+        IStorageUtilityIndexed<FormRecord> formRecordstorage =
+                CommCareApplication.instance().getUserStorage(FormRecord.class);
+        formRecordstorage.write(r);
+    }
+
+
+    private static class InvalidStateException extends Exception {
+        public InvalidStateException(String message) {
+            super(message);
+        }
+    }
+
+    private static SqlStorage<InstanceRecord> getInstanceRecordStorage() {
+        if (sInstanceRecordStorage == null) {
+            sInstanceRecordStorage = CommCareApplication.instance().getAppStorage(InstanceRecord.class);
+        }
+        return sInstanceRecordStorage;
+    }
+
+    public static void setinstanceRecordStorage(SqlStorage<InstanceRecord> instanceRecordStorage) {
```

Comment:
I'm not OK with managing this state statically. I know it's annoying, but I think we need to refactor to move all of the static functionality in this class to being called through some other pattern. 

It would be OK in my mind for it to be linked through the platform through some context object, but it has only ever caused a lot of headache/tech debt to allow static maintenance of the storage sandboxes, so we should figure out how to make this something that comes from the platform.

Reply:
will do, though I would be curious to know what kind of problems have we seen in past arising from storage sadboxes being singletons.

Reply:
Two big things that have come up a lot:

1) We already have a *lot* of lifecycle objects around the code (java lifecycle, app lifecycle, user lifecycle) and it's a lot of overhead to maintain registering and unregistering data from them.

2) In many cases we are on a track to being able to decouple static state from apps entirely, which would give us pretty incredible capabilities like the ability to update apps in the background without interfering with seated applications. Adding new statically managed holders (rather than coupling them into another lifecycle object) significantly complicates those efforts by making a full refactor necessary and requiring us to inject the new lifecycle object in lots of places (this came up a ton with things like ReferenceManager, and it's a pattern we want to move away from)

Reply:
Thanks! Currently we have different versions of `getUserStorage` and `getAppStorage` in Application class [here](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/CommCareApplication.java#L556). I am thinking of using these same functions for FormDefRecord and InstanceRecord to replace the static counterparts. Does that sound right or are we looking for something alternate/better here ?

Reply:
That should be fine. Passing the storage as a constructor to this (and/or wrapping that operation in a static getter from the Application) is a fine solution, since that puts us in the same state.

As a best-practice pattern I try to only call the singleton application methods from lifecycle objects (like the Activity classes), because it makes it easier to manage the outcome if the lifecycle is in a bad state (say, the user has been logged out).